### ISSUE TITLE
regionprops hodge-podge of output types - optionally return pd.DataFrame ?

### ISSUE THREAD

**Issue by VolkerH** [scikit-image/scikit-image#3180]
## Description

Currently, `regionprops` provides a list of `skimage.measure._regionprops._RegionProperties`.
These in turn contain the properties,  some of which are tuples of integers, others are `np.ndarrays` and others or scalar types. 

I would argue that a common use case for `regionprops` is to calculate features for subsequent classification of these image regions. For this it would be most useful to have the numerical features in a tabular form, ideally a data frame where the rows correspond to the object labels and the columns correspond to the features. For this, some of the features (tuples, ndarrays) such as for example the moments must be flattened. 
Other features such as `coords` are probably not too useful in this scenario.
  
Currently, I am using some fairly ugly code to loop over the regions, loop over all features, checking each one against a list of which features I want to keep and flattening them according to their data type. 

I assume returning a data frame directly from `regionprops` is not an option as it would introduce a dependency on pandas and would break backward compatibility. But given that pandas seems to be here to stay such a dependency could be acceptable. 
I would argue that a helper function as part of `skimage` that takes the output of regionprops and turns it into a `pd.DataFrame` would be highly useful. This would also adress the issue raised in https://github.com/scikit-image/scikit-image/issues/2755

**Comment by stefanv** [comment]
+1 on a way to get regionprops mapped back to the original image, or to a table.  Both would be useful.

**Comment by VolkerH** [comment]
@jni I agree a numpy array plus a list of column names would be fully sufficient and trivial to convert into a pandas dataframe.
In that case, do you propose to cast all features (many of which are integers) to `np.float` for the final table?

**Comment by jni** [comment]
> do you propose to cast all features (many of which are integers) to np.float for the final table

Yes.

**Comment by jni** [comment]
@soupault yes, but they can’t be used as input to sklearn.

**Comment by kmader** [comment]
great idea, and I agree skimage shouldn't add pandas as a hard dependency. However, as someone who has taught a number of introductory courses to image analysis, I feel like the typical user would appreciate having a meaningful html/pandas rendered view when they run regionprops. DataFrame does this quite well and so maybe having it use data.frame if pandas is available and fallback to dicts or arrays+columns if not and having an argument `use_dataframe` to allow for more predictible behavior.

I've done some work like this before in case someone is interested (it isn't elegant but it works, https://github.com/4Quant/pyqae/blob/59f43786520c3a216a8e6b79ce6e73f8f5933448/pyqae/shape_analysis.py#L27-L70)

**Comment by stefanv** [comment]
I would be -1 on the output types changing depending on the system configuration, but if we return a dictionary it makes it quite easy to construct a dataframe from it (`pandas.DataFrame.from_dict`), and we can document this too.

Having `pandas` as an optional dependency just for this shouldn't be a problem.  We can add both `regionprops_as_df` and `regionprops_as_dict`, and re-use the underlying implementation.

**Comment by jni** [comment]
Another issue as I start to think about this more closely: many region properties are multi-dimensional. I think the "correct" way to deal with these is as a multi-index for pandas columns. However the multi-indexing doesn't work great when mixing scalar values and multi-values:

```python
In [16]: x = pd.DataFrame({
    ...:     ('foo', None): [5, 6, 7],
    ...:     ('bar', 1): [8., 9.8, 2.2],
    ...:     ('bar', 2): [9., 2., 2.4],
    ...: })

In [17]: x
Out[17]: 
  foo  bar     
  NaN    1    2
0   5  8.0  9.0
1   6  9.8  2.0
2   7  2.2  2.4
```

Options:

- Use hierarchical (tuple) indices with three levels: feature name, index for axis 0, index for axis 1. This lets us represent features up to 2D arrays. I think that encompasses all features except `image` and `label_image`. Where we only have 1D or scalars, we use 0 for the first level or both levels, respectively.
- Use hierarchical (tuple) indices with two levels: feature name, and linear (raveled) index. This works but you lose information about feature shape. Maybe this is fine anyway — if you need round-trip accuracy then you should build your own wrapper around the current regionprops?
- Use a single level index, where we concatenate the multiple levels above where present. e.g. `'inertia_tensor-1-2'`.
- Use object columns. This is my least preferred method but the most generic.

@scikit-image/core @VolkerH any opinions here?

@kmader if I understand your code, you just discarded higher-dimensional features. What do you think of the above options?

**Comment by VolkerH** [comment]
Tricky one. At this stage, you probably want to maintain backwards compatibility at least by enabling an option.

I would prefer to have the single level index, with the string giving hints about dimensionality, e.g. the `inertia_tensor-1-2` or something like `moments_0_0`. Maybe there could even be a naming scheme such as 

`scalar_name` with `name` being the feature name, e.g. mean_intensity
`vector_name_n` (with `n` being an integer)
`mat_name_n_m`
I guess putting the type information in the string is not very clean, but it would allow for fairly easy filtering by the users to narrow it down to the type of features they want.

As you mention `image` and `label_image` are special cases. I have used those to create thumbnails, but I cannot think of lots of use cases where one would want to unravel those into a feature table. 
There should be an easy way to drop these (maybe a simple flag passed as a function argument).

**Comment by kmader** [comment]
The `scalar_`,  `vector_`, and  `mat_` seems like a very nice solution for most of the use cases I have seen. It then let's you use pandas to easily filter out the columns you want to keep.
It lets you directly use sklearn and also is very easy to save to disk as csv or json which can get tricked up when values are anything but scalars.

**Comment by jni** [comment]
Thanks for your comments! Some further thoughts:

- the separator should definitely be hyphen or spaces or something else that is not an underscore, since some feature names have underscores (e.g. `inertia_tensor`).
- I suggest not including `vec` etc in the name. `startswith('vec')` is easily replaced with `len(elem.split('-')) == 1`. Maybe the pandas filtering is marginally more cumbersome but I think that's a worthwhile tradeoff.
- backwards compatibility is a major concern and my proposed solution currently is to make a new function, `regionprops_table(label_image, intensity_image, *, property_list=None)`. That will either return a format compatible with `pd.DataFrame` so that it's trivial to wrap, *or* an option to specify dataframe as a return type. I prefer the former because it's a bit icky to have different return types based on input values. We could add the function `regionprops_dataframe`, also, that is only implemented if pandas is installed. But there should be some real consistency between all the options.

**Comment by kmader** [comment]
Those suggestions make sense, 
- As for spaces/hyphens I would tend to go with hyphens as `inertial_tensor 2 3 1` seems less intuitive for me than `intertial_tensor-2-3-1`
-  I can now see a few cases where having the `mat_`, `vec_` prefixes might be problematic because in 2D they are vectors and in 3D they could be matrices and so columns with changing prefixes might be annoying. 
- `pd.DataFrame(regionprops_table(label_image, intensity_image))` is certainly easy enough for students to learn. I also agree that having output types change based on input arguments is not preferable and particularly as more teams are using mypy and type annotations this makes life much more difficult for them. I don't think a `regionprops_dataframe` function is that important and having a function that only works when an optional dependency is installed could be confusing.

**Comment by VolkerH** [comment]
Fully agree with @kmader here, lot's of good points that I hadn't thought about.
